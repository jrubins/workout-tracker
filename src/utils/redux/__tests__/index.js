import { createDataReducer } from '../'

describe('redux tests', () => {
  describe('createDataReducer', () => {
    const types = {
      fetch: 'fetch',
      fetchFn: 'fetchFn',
      fetchWithoutWrapper: 'fetchWithoutWrapper',
    }
    const saveActionTypes = {
      [types.fetch]: 'things',
      [types.fetchWithoutWrapper]: '',
      [types.fetchFn]: data => data.map(({ nestedThing }) => nestedThing),
    }
    const dataPoint = {
      foo: 'bar',
      id: 1,
    }

    const dataReducer = createDataReducer({
      saveActionTypes,
    })

    test('returns data on successfully fetch', () => {
      expect(
        dataReducer([], {
          data: {
            [saveActionTypes[types.fetch]]: [dataPoint],
          },
          type: types.fetch,
        })
      ).toMatchSnapshot()
    })

    test('returns data on successfully fetchWithoutWrapper', () => {
      expect(
        dataReducer([], {
          data: [dataPoint],
          type: types.fetchWithoutWrapper,
        })
      ).toMatchSnapshot()
    })

    test('uniquely saves data by ID', () => {
      expect(
        dataReducer([], {
          data: {
            [saveActionTypes[types.fetch]]: [dataPoint, dataPoint],
          },
          type: types.fetch,
        })
      ).toMatchSnapshot()
    })

    test('prefers newer data over old data', () => {
      expect(
        dataReducer([], {
          data: {
            [saveActionTypes[types.fetch]]: [
              {
                ...dataPoint,
                newDataPoint: 'newData',
              },
              dataPoint,
            ],
          },
          type: types.fetch,
        })
      ).toMatchSnapshot()
    })

    test('allows a function to be used in the saveActionTypes keys', () => {
      expect(
        dataReducer([], {
          data: [
            {
              nestedThing: {
                blah: 'kazaam',
                id: 1,
              },
            },
            {
              nestedThing: {
                huey: 'kaphooey',
                id: 2,
              },
            },
          ],
          type: types.fetchFn,
        })
      ).toMatchSnapshot()
    })

    test('allows an optional data transformer to be specified', () => {
      const transformerDataReducer = createDataReducer({
        dataTransformerFn: data =>
          data.map(dataPoint => ({
            ...dataPoint,
            customAttr: 'fakestuff',
          })),
        saveActionTypes,
      })
      expect(
        transformerDataReducer([], {
          data: {
            [saveActionTypes[types.fetch]]: [dataPoint],
          },
          type: types.fetch,
        })
      ).toMatchSnapshot()
    })

    test('deletes data', () => {
      const deleteActionType = 'deleteActionType'
      const deleteDataReducer = createDataReducer({
        deleteActionTypes: [deleteActionType],
      })
      expect(
        deleteDataReducer([dataPoint], {
          id: dataPoint.id,
          type: deleteActionType,
        })
      ).toMatchSnapshot()
    })

    test('clears data', () => {
      const clearActionType = 'clearActionType'
      const clearDataReducer = createDataReducer({
        clearActionTypes: [clearActionType],
      })
      expect(
        clearDataReducer([dataPoint], {
          type: clearActionType,
        })
      ).toMatchSnapshot()
    })
  })
})
