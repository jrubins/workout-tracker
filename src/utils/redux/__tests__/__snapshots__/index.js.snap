// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux tests createDataReducer allows a function to be used in the saveActionTypes keys 1`] = `
Array [
  Object {
    "blah": "kazaam",
    "id": 1,
  },
  Object {
    "huey": "kaphooey",
    "id": 2,
  },
]
`;

exports[`redux tests createDataReducer allows an optional data transformer to be specified 1`] = `
Array [
  Object {
    "customAttr": "fakestuff",
    "foo": "bar",
    "id": 1,
  },
]
`;

exports[`redux tests createDataReducer clears data 1`] = `Array []`;

exports[`redux tests createDataReducer deletes data 1`] = `Array []`;

exports[`redux tests createDataReducer prefers newer data over old data 1`] = `
Array [
  Object {
    "foo": "bar",
    "id": 1,
    "newDataPoint": "newData",
  },
]
`;

exports[`redux tests createDataReducer returns data on successfully fetch 1`] = `
Array [
  Object {
    "foo": "bar",
    "id": 1,
  },
]
`;

exports[`redux tests createDataReducer returns data on successfully fetchWithoutWrapper 1`] = `
Array [
  Object {
    "foo": "bar",
    "id": 1,
  },
]
`;

exports[`redux tests createDataReducer uniquely saves data by ID 1`] = `
Array [
  Object {
    "foo": "bar",
    "id": 1,
  },
]
`;
